src/codeflare_sdk/jobs/jobs.py

```diff
@@ -0,0 +1,162 @@
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+The jobs sub-module contains the definitions for the Job objects, which represent
+the methods by which a user can submit a job to a cluster, check the jobs status and 
+access the jobs logs. 
+"""
+
+import abc
+from typing import List
+from pathlib import Path
+
+from ray.job_submission import JobSubmissionClient
+from torchx.components.dist import ddp
+from torchx.runner import get_runner, Runner
+from torchx.specs import AppHandle, parse_app_handle, AppDryRunInfo
+
+from .config import JobConfiguration
+
+import typing
+
+if typing.TYPE_CHECKING:
+    from ..cluster.cluster import Cluster
+
+all_jobs: List["Job"] = []
+torchx_runner: Runner = get_runner()
+
+
+class JobDefinition(metaclass=abc.ABCMeta):
+    """
+    A job definition to be submitted to a generic backend cluster.
+    """
+
+    def _dry_run(self, cluster) -> str:
+        """
+        Create job definition, but do not submit.
+
+        The primary purpose of this function is to facilitate unit testing.
+        """
+
+    def submit(self, cluster: "Cluster"):
+        """
+        Method for creating a job on a specific cluster
+        """
+        pass
+
+
+class Job(metaclass=abc.ABCMeta):
+    """
+    An abstract class that defines the necessary methods for authenticating to a remote environment.
+    Specifically, this class `status` and a `logs` function.
+    """
+
+    def status(self):
+        """
+        Method for retrieving the job's current status.
+        """
+        pass
+
+    def logs(self):
+        """
+        Method for retrieving the job's logs.
+        """
+
+
+class TorchXJobDefinition(JobDefinition):
+    def __init__(self, config: JobConfiguration):
+        """
+        Create the TorchXJob object by passing in a JobConfiguration
+        (defined in the config sub-module).
+        """
+        self.config = config
+
+    def _dry_run(self, cluster: "Cluster", *script_args) -> AppDryRunInfo:
+        """
+        Create job definition, but do not submit.
+
+        The primary purpose of this function is to facilitate unit testing.
+        """
+        j = f"{cluster.config.max_worker}x{max(cluster.config.gpu, 1)}"  # # of proc. = # of gpus
+        dashboard_address = f"{cluster.cluster_dashboard_uri(cluster.config.namespace).lstrip('http://')}"
+        return torchx_runner.dryrun(
+            app=ddp(
+                *script_args,
+                script=self.config.script,
+                m=self.config.m,
+                name=self.config.name,
+                h=None,  # for custom resource types
+                cpu=cluster.config.max_cpus,
+                gpu=cluster.config.gpu,
```

Comment:
Can we have these be the defaults, but expose them as options to the user. This setup assume you want to use 100% of cluster resources for every job submitted. Not sure if this would always be the case.